### PR TITLE
Direct allocation attribution in profiling

### PR DESCRIPTION
**Pull request by yinan1048576**
This change is not completed yet, but I think it has reached a stage suitable for seeking some early feedback.

The context is in #1751. To enable correct inference, we need to record both the number of sampled bytes and the number of unsampled bytes for each stack trace. The number of sampled bytes is equivalent to `curobjs`, so we only need to record the number of unsampled bytes, which I denote as `surplus` - the amount beyond what's needed for triggering sampling.

`surplus` is relayed in two routes:
- `thread_alloc_event()` -> `prof_malloc_sample_object()`, via a new thread local field `prof_sample_event_surplus`, which is needed because the sampling logic is currently separate from the thread event logic;
- `prof_malloc_sample_object()` -> `prof_free_sampled_object()`, via a new field `e_prof_surplus` on `edata`, which is needed because `free()` needs to roll back the allocation's `surplus` from the stack trace total.

My remaining work involves the following:
- Add unit tests.
- Add an option to actually use the additional information i.e. the `surplus`. (I figured that computing the `surplus` is fairly cheap so I don't really need to guard the computation under any option.)
  - Print the raw `surplus` value as well as some other convenient estimation(s) to the profiling dump.
  - Compute leak checking differently.

### CODE REVIEW

**Comment by yinan1048576** [comment]
Rebase, and rewrite. The main change in the rewrite is that I'm not creating the new thread local field any more, thanks to #1779.

File: include/jemalloc/internal/thread_event.h [hunk 1]
```diff
@@ -217,12 +217,44 @@ te_ctx_get(tsd_t *tsd, te_ctx_t *ctx, bool is_alloc) {
...
+
+JEMALLOC_ALWAYS_INLINE bool
+te_prof_sample_event_lookahead_surplus(tsd_t *tsd, size_t usize,
+    size_t *surplus) {
+	/* The subtraction is intentionally susceptible to underflow. */
+	size_t accumbytes = tsd_thread_allocated_get(tsd) + usize -
+	    tsd_thread_allocated_last_event_get(tsd);
+	size_t sample_wait = tsd_prof_sample_event_wait_get(tsd);
+	if (accumbytes < sample_wait) {
+		/* Don't bother to set surplus - it will never be read. */
```
Comment (yinan1048576) [inline include/jemalloc/internal/thread_event.h]: This may trigger some warning. Let's see...

Comment (yinan1048576) [reply include/jemalloc/internal/thread_event.h]: As expected. Adding an initializer.

**Comment by yinan1048576** [comment]
Fix compiler warnings, as well as one other place revealed by the FreeBSD unit tests: `te_prof_sample_event_lookahead_surplus()` would better be short-circuited in edge cases e.g. non-nominal TSD / reentrancy.

**Comment by yinan1048576** [comment]
Rebase on top of #1787 to get rid of the warning.

File: include/jemalloc/internal/arena_inlines_b.h [hunk 2]
```diff
@@ -101,11 +101,11 @@ arena_prof_tctx_reset_sampled(tsd_t *tsd, const void *ptr) {
 }
 
 JEMALLOC_ALWAYS_INLINE void
-arena_prof_info_set(edata_t *edata, prof_tctx_t *tctx) {
+arena_prof_info_set(edata_t *edata, prof_tctx_t *tctx, uint64_t surplus) {
```
Comment (davidtgoldblatt) [inline include/jemalloc/internal/arena_inlines_b.h]: Surplus should be a size_t, no?

File: include/jemalloc/internal/thread_event.h [hunk 3]
```diff
@@ -217,12 +217,56 @@ te_ctx_get(tsd_t *tsd, te_ctx_t *ctx, bool is_alloc) {
...
+    size_t *surplus) {
+	if (surplus != NULL) {
+		/*
+		 * This is a dead store: the surplus will be overwritten before
+		 * any read.  The initialization suppresses compiler warnings.
+		 * Meanwhile, using usize to initialize is good for debugging
+		 * purpose, because a valid surplus value is strictly less than
+		 * usize.
+		 */
+		*surplus = usize;
```
Comment (davidtgoldblatt) [inline include/jemalloc/internal/thread_event.h]: If I might suggest: I'm not sure I would spot that `*surplus == usize` is invalid while debugging. Some other garbage constant (e.g. `0x99999999` or `0xdeadbeef` or something) OTOH, I would.

Comment (yinan1048576) [reply include/jemalloc/internal/thread_event.h]: Makes sense. Use `SIZE_MAX` instead. (Strictly speaking `0x99999999` and `0xdeadbeef` can be valid `surplus` values.)

**Comment by yinan1048576** [comment]
Revise according to feedback.

**Comment by yinan1048576** [comment]
Rebase, and added another field `accumsurplus`. It's analogous to `cursurplus`, but records the accumulative surplus, and is computed whenever `opt_prof_accum` is on.

My next step is to define an opt-in run-time option and, when it is turned on, to ingest the surplus values in the dump content.

**Comment by yinan1048576** [comment]
Rebase.

**Comment by yinan1048576** [comment]
Rebase, and added:
- Unit tests.
- An option to print total surplus rather than total bytes in prof dumps. Users don't need the total bytes and can derive whatever inference they need from
  - the number of sampled objects,
  - total surplus, and
  - `lg_prof_sample`.

**Comment by yinan1048576** [comment]
Closing. #1897 serves to correct the bias. Will reopen if there's any need for inference, in addition to a single point estimate.
